src/drawPath.cpp

```diff
@@ -0,0 +1,243 @@
+{
+	auto status = psDroid->sMove.Status;
+	if (status == MOVEINACTIVE || status == MOVEPAUSE || status == MOVEWAITROUTE || psDroid->sMove.asPath.empty())
+		return;
+
+	const int terrainOffset = 5;
+
+	std::vector<Vector3i> pastPoints, futurePoints;
+
+	Vector3i prev;
+	int currentPt = psDroid->sMove.pathIndex;
+	for (int i = 0; i < psDroid->sMove.asPath.size(); i++) {
+		Vector2i pt2 = psDroid->sMove.asPath[i];
+		auto height = map_TileHeight(map_coord(pt2.x), map_coord(pt2.y)) + terrainOffset;
+		Vector3i pt3(pt2.x, height, -pt2.y);
+
+		if (i > 0) {
+			if (i < currentPt) {
+				pastPoints.push_back(prev);
+				pastPoints.push_back(pt3);
+			} else if (i >= currentPt) {
+				futurePoints.push_back(prev);
+				futurePoints.push_back(pt3);
+			}
+		}
+		prev = pt3;
+	}
+
+	if (futurePoints.size() > 1)
+		drawLines (perspectiveViewMatrix, futurePoints, WZCOL_GREEN);
+	if (pastPoints.size() > 1)
+		drawLines (perspectiveViewMatrix, pastPoints, WZCOL_YELLOW);
+
+	auto height = map_TileHeight(map_coord(psDroid->sMove.target.x), map_coord(psDroid->sMove.target.y)) + terrainOffset;
+	drawSquare(perspectiveViewMatrix, 32, psDroid->sMove.target.x, psDroid->sMove.target.y, height, WZCOL_GREEN);
+}
+
+/// Draws path tree from the context.
+void drawContext(const PathfindContext& context, const glm::mat4 &mvp)
+{
+	const int terrainOffset = 5;
+	const int contextIteration = context.iteration;
+	std::vector<Vector3i> pts;
+	for (int y = 0; y < context.height; y++)
+	{
+		for (int x = 0; x < context.width; x++)
+		{
+			const PathExploredTile& tile = context.tile(PathCoord(x, y));
+			if (tile.iteration != contextIteration)
+				continue;
+			auto heightStart = map_TileHeight(x, y) + terrainOffset;
+			Vector3i start(world_coord(x), heightStart, -world_coord(y));
+			pts.push_back(start);
+			auto heightEnd = map_TileHeight(x - tile.dx/64, y - tile.dy/64) + terrainOffset;
+			Vector3i end(world_coord(x) - tile.dx , heightEnd, -world_coord(y) + tile.dy);
+			pts.push_back(end);
+		}
+	}
+	if (!pts.empty())
+		drawLines(mvp, pts, WZCOL_LGREEN);
+}
+
+static bool _drawImpassableTiles = false;
+static bool _drawContextTree = false;
+static bool _drawDroidPath = false;
+
+extern std::list<PathfindContext> fpathContexts;
+
+void drawPathCostLayer(int player, const iView& playerViewPos, const glm::mat4& viewMatrix, const glm::mat4 &perspectiveViewMatrix)
+{
+	if (player >= MAX_PLAYERS)
+		return;
+
+	if (_drawImpassableTiles) {
+		PathBlockingMap bmap;
+		PathBlockingType pbtype;
+		pbtype.owner = player;
+		pbtype.propulsion = PROPULSION_TYPE_WHEELED;
+		fillBlockingMap(bmap, pbtype);
+		debugDrawImpassableTiles(bmap, playerViewPos, perspectiveViewMatrix, 6);
+	}
+
+	if (_drawDroidPath)
+	{
+		for (DROID *psDroid = apsDroidLists[player]; psDroid; psDroid = psDroid->psNext)
+		{
+			if (psDroid->selected)
+			{
+				drawDroidPath(psDroid, viewMatrix, perspectiveViewMatrix);
+			}
+		}
+	}
+
+	if (_drawContextTree && !fpathContexts.empty())
```

Comment:
Thread-safe? (i.e. when `fpathAsyncMode = true`)

(We can just condition this on `fpathAsyncMode`, and if someone wants to draw the context tree for debugging they have to modify `fpathAsyncMode` to false.)

Reply:
Oops. Seems like the old rendering of paths was not thread-safe as well. Maybe it is reasonable not to draw debug paths in threaded mode for now. 

Though I am considering a more robust mechanism for transferring tasks to PF threads and gathering results back.

Reply:
Done in simplest way